src/uu/cp/src/cp.rs

```diff
@@ -866,6 +871,47 @@ fn copy_source(
     }
 }
 
+pub fn normalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
+    let mut components = path.as_ref().components().peekable();
+    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+        let buf = PathBuf::from(c.as_os_str());
+        components.next();
+        buf
+    } else {
+        PathBuf::new()
+    };
+
+    for component in components {
+        match component {
+            Component::Prefix(..) => unreachable!(),
+            Component::RootDir => {
+                ret.push(component.as_os_str());
+            }
+            Component::CurDir => {}
+            Component::ParentDir => {
+                ret.pop();
+            }
+            Component::Normal(c) => {
+                ret.push(c);
+            }
+        }
+    }
+    ret
+}
+
+#[cfg(target_os = "windows")]
+fn adjust_canonicalization<'a, P: 'a + AsRef<Path> + Into<PathBuf>>(p: P) -> Cow<'a, Path> {
```

Comment:
After looking at how you use this function in context, I think you can change it to:
```rust
#[cfg(target_os = “windows”)]
fn adjust_canonicalization<'a>(p: &'a Path) -> Cow<'a, Path> {
    // In some cases, \\? can be missing on some Windows paths.  Add it at the
    // beginning unless the path is prefixed with a device namespace.
    const VERBATIM_PREFIX: &str = r#"\\?"#;
    const DEVICE_NS_PREFIX: &str = r#"\\."#;

    let has_prefix = p
        .components()
        .next()
        .and_then(|comp| comp.as_os_str().to_str())
        .map(|p_str|
            p_str.starts_with(VERBATIM_PREFIX)
                || p_str.starts_with(DEVICE_NS_PREFIX))
        .unwrap_or_default();

    if has_prefix {
        p.into()
    } else {
        Path::new(VERBATIM_PREFIX).join(p).into()
    }
}
```

This avoids the unnecessary allocation and also only converts the first component into an `&str`.  It also checks for `\\.`, which I do not believe can have a verbatim prefix (which seems to be the case at least based on the Rust docs).

Reply:
thanks!